RedHatInsights/rhsm-subscriptions
# Swatch-1691: event kafka retry

Jira issue: [SWATCH-1691](https://issues.redhat.com/browse/SWATCH-1691)

## Description

In order to handle failed Event kafka messages we need to retry and then if failing continues put message on a dead letter queue to be processed later.

We only put messages with valid json that can be parsed into an Event object but not able to be persisted on the dead letter topic.

Topic creation (topic had to be renamed to all lowercase): https://github.com/RedHatInsights/platform-mq/pull/305 and https://github.com/RedHatInsights/platform-mq/pull/307
## Testing

### Setup

1. Clean up the Events table.
2. Save event that will be duplicated in messages and should fail on message consumption:
`INSERT INTO public.events VALUES ('b1208963-b92a-4c59-a730-88fd5fadb9b1', NULL, '2023-05-02 00:00:00+00', '{"sla": "Premium", "role": "osd", "org_id": "146", "event_id": "b1208963-b92a-4c59-a730-88fd5fadb9b1", "timestamp": "2023-05-02T00:00:00Z", "event_type": "snapshot_redhat.com:openshift_dedicated:cluster_hour", "expiration": "2023-05-02T01:00:00Z", "instance_id": "e3a62bd1-fd00-405c-9401-f2288808588d", "display_name": "automation_osd_cluster_e3a62bd1-fd00-405c-9401-f2288808588d", "event_source": "prometheus", "measurements": [{"uom": "Instance-hours", "value": 1.0}], "service_type": "OpenShift Cluster"}', 'snapshot_redhat.com:openshift_dedicated:cluster_hour', 'prometheus', 'e3a62bd1-fd00-405c-9401-f2288808588d', '146', DEFAULT, NULL, '2023-09-26 14:57:56.160104+00');`
3. Use the Big data Tools of Intellij as producer to test the consumer topic `platform.rhsm-subscriptions.service-instance-ingress`. Payloads are as below make sure to check each Test step on which payload to use:
![image](https://github.com/RedHatInsights/rhsm-subscriptions/assets/11471167/03ab9bbf-f1a9-49a9-a980-4fb3947dce6d)

Payload144: `{"sla": "Premium", "role": "osd", "org_id": "144","timestamp": "2023-05-02T00:00:00Z", "event_type": "snapshot_redhat.com:openshift_dedicated:cluster_hour", "expiration": "2023-05-02T01:00:00Z", "instance_id": "e3a62bd1-fd00-405c-9401-f2288808588d", "display_name": "automation_osd_cluster_e3a62bd1-fd00-405c-9401-f2288808588d", "event_source": "prometheus", "measurements": [{"uom": "Instance-hours", "value": 1.0}], "service_type": "OpenShift Cluster"}`

Payload145: `{"sla": "Premium1", "role": "osd", "org_id": "145","timestamp": "2023-05-02T00:00:00Z", "event_type": "snapshot_redhat.com:openshift_dedicated:cluster_hour", "expiration": "2023-05-02T01:00:00Z", "instance_id": "e3a62bd1-fd00-405c-9401-f2288808588d", "display_name": "automation_osd_cluster_e3a62bd1-fd00-405c-9401-f2288808588d", "event_source": "prometheus", "measurements": [{"uom": "Instance-hours", "value": 1.0}], "service_type": "OpenShift Cluster"}`

Payload146: `{"sla": "Premium", "role": "osd", "org_id": "146","timestamp": "2023-05-02T00:00:00Z", "event_type": "snapshot_redhat.com:openshift_dedicated:cluster_hour", "expiration": "2023-05-02T01:00:00Z", "instance_id": "e3a62bd1-fd00-405c-9401-f2288808588d", "display_name": "automation_osd_cluster_e3a62bd1-fd00-405c-9401-f2288808588d", "event_source": "prometheus", "measurements": [{"uom": "Instance-hours", "value": 1.0}], "service_type": "OpenShift Cluster"}`

### Steps

1. **Step 1: Add new events**
In the value string add payload 144, payload 145, payload 146 one by one in order and produce.
2. Start app with: `DEV_MODE=true ./gradlew clean :bootRun`

### Verification

1. One new entry in events table should exist for org_id 144 `select * from events where org_id = 144`
2. Check logs for `[INFO ] [org.candlepin.subscriptions.tally.billing.ServiceInstanceMessageConsumer] - Events received w/ event list size=3. Consuming events.`
3. Should error out and retry
4. You should also see a log where we get an event list size=1 since that is the org_id 146 record being retried since it failed to save.
5. Use big data tool to consume messages from platform.rhsm-subscriptions.service-instance-ingress.DLT which should contain the org_id 146 event.
